File: circularimageview/src/main/java/com/github/abdularis/civ/AvatarImageView.java
package com.github.abdularis.civ;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.RectF;
import android.util.AttributeSet;

import androidx.annotation.ColorInt;
import androidx.annotation.Dimension;
import androidx.annotation.IntDef;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

/**
 * Created by abdularis on 22/03/18.
 */

public class AvatarImageView extends CircleImageView {

    public static final int SHOW_INITIAL = 1;
    public static final int SHOW_IMAGE = 2;

    @IntDef({SHOW_INITIAL, SHOW_IMAGE})
    @Retention(RetentionPolicy.SOURCE)
    public @interface State {
    }

    private static final String DEF_INITIAL = "A";
    private static final int DEF_TEXT_SIZE = 90;
    private static final int DEF_BACKGROUND_COLOR = 0xE53935;
    @State
    private static final int DEF_STATE = SHOW_INITIAL;

    private Paint mTextPaint;
    private Rect mTextBounds;

    private Paint mBackgroundPaint;
    private RectF mBackgroundBounds;

    @NonNull
    private String mInitial;
    @NonNull
    private String mText;

    private int mShowState;

    public AvatarImageView(Context context) {
        this(context, null);
    }

    public AvatarImageView(Context context, AttributeSet attrs) {
        super(context, attrs);

        String text = DEF_INITIAL;
        int textColor = Color.WHITE;
        int textSize = DEF_TEXT_SIZE;
        int backgroundColor = DEF_BACKGROUND_COLOR;
        int showState = DEF_STATE;

        if (attrs != null) {
            TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.AvatarImageView, 0, 0);

            text = a.getString(R.styleable.AvatarImageView_text);
            textColor = a.getColor(R.styleable.AvatarImageView_textColor, textColor);
            textSize = a.getDimensionPixelSize(R.styleable.AvatarImageView_textSize, textSize);
            backgroundColor = a.getColor(R.styleable.AvatarImageView_avatarBackgroundColor, backgroundColor);
            showState = a.getInt(R.styleable.AvatarImageView_view_state, showState);

            a.recycle();
        }

        mShowState = showState;
        mTextPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mTextPaint.setTextAlign(Paint.Align.CENTER);
        mTextPaint.setColor(textColor);
        mTextPaint.setTextSize(textSize);

        mTextBounds = new Rect();
        mText = text == null ? "" : text;
        mInitial = extractInitial(text);
        updateTextBounds();

        mBackgroundPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mBackgroundPaint.setColor(backgroundColor);
        mBackgroundPaint.setStyle(Paint.Style.FILL);

        mBackgroundBounds = new RectF();
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        updateCircleDrawBounds(mBackgroundBounds);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        if (mShowState == SHOW_INITIAL) {
            float textBottom = mBackgroundBounds.centerY() - mTextBounds.exactCenterY();
            canvas.drawOval(mBackgroundBounds, mBackgroundPaint);
            canvas.drawText(mInitial, mBackgroundBounds.centerX(), textBottom, mTextPaint);
            drawStroke(canvas);
            drawHighlight(canvas);
        } else {
            super.onDraw(canvas);
        }
    }

    @NonNull
    public String getInitial() {
        return mInitial;
    }

    @NonNull
    public String getText() {
        return mText;
    }

    public void setText(@Nullable String text) {
        mText = text == null ? "" : text;
        mInitial = extractInitial(text);
        updateTextBounds();
        invalidate();
    }

    @State
    public int getState() {
        return mShowState;
    }

    public void setState(@State int state) {
        if (state != SHOW_INITIAL && state != SHOW_IMAGE) {
            String msg = "Illegal avatar state value: " + state + ", use either SHOW_INITIAL or SHOW_IMAGE constant";
            throw new IllegalArgumentException(msg);
        }
        mShowState = state;
        invalidate();
    }

    @Dimension
    public float getTextSize() {
        return mTextPaint.getTextSize();
    }

    public void setTextSize(@Dimension float size) {
        mTextPaint.setTextSize(size);
        updateTextBounds();
        invalidate();
    }

    @ColorInt
    public int getTextColor() {
        return mTextPaint.getColor();
    }

    public void setTextColor(@ColorInt int color) {
        mTextPaint.setColor(color);
        invalidate();
    }

    @ColorInt
    public int getAvatarBackgroundColor() {
        return mBackgroundPaint.getColor();
    }

    public void setAvatarBackgroundColor(@ColorInt int color) {
        mBackgroundPaint.setColor(color);
        invalidate();
    }

    @NonNull
    private String extractInitial(@Nullable String letter) {
        if (letter == null || letter.trim().length() <= 0) return "?";
        return String.valueOf(letter.charAt(0));
    }

    private void updateTextBounds() {
        mTextPaint.getTextBounds(mInitial, 0, mInitial.length(), mTextBounds);
    }
}

File: circularimageview/src/main/java/com/github/abdularis/civ/AvatarImageView.kt
package com.github.abdularis.civ

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import androidx.annotation.ColorInt
import androidx.annotation.Dimension
import androidx.annotation.IntDef

/**
 * Created by abdularis on 22/03/18.
 */
class AvatarImageView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null
) : CircleImageView(context, attrs) {
    @IntDef(SHOW_INITIAL, SHOW_IMAGE)
    annotation class State

    private val mTextPaint: Paint
    private val mTextBounds: Rect
    private val mBackgroundPaint: Paint
    private val mBackgroundBounds: RectF
    private var initial: String
    private var mText: String
    private var mShowState: Int
    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        updateCircleDrawBounds(mBackgroundBounds)
    }

    override fun onDraw(canvas: Canvas) {
        if (mShowState == SHOW_INITIAL) {
            val textBottom = mBackgroundBounds.centerY() - mTextBounds.exactCenterY()
            canvas.drawOval(mBackgroundBounds, mBackgroundPaint)
            canvas.drawText(initial, mBackgroundBounds.centerX(), textBottom, mTextPaint)
            drawStroke(canvas)
            drawHighlight(canvas)
        } else {
            super.onDraw(canvas)
        }
    }

    var text: String?
        get() = mText
        set(text) {
            mText = text ?: ""
            initial = extractInitial(text)
            updateTextBounds()
            invalidate()
        }

    @get:State
    var state: Int
        get() = mShowState
        set(state) {
            if (state != SHOW_INITIAL && state != SHOW_IMAGE) {
                val msg =
                    "Illegal avatar state value: $state, use either SHOW_INITIAL or SHOW_IMAGE constant"
                throw IllegalArgumentException(msg)
            }
            mShowState = state
            invalidate()
        }

    @get:Dimension
    var textSize: Float
        get() = mTextPaint.textSize
        set(size) {
            mTextPaint.textSize = size
            updateTextBounds()
            invalidate()
        }

    @get:ColorInt
    var textColor: Int
        get() = mTextPaint.color
        set(color) {
            mTextPaint.color = color
            invalidate()
        }

    @get:ColorInt
    var avatarBackgroundColor: Int
        get() = mBackgroundPaint.color
        set(color) {
            mBackgroundPaint.color = color
            invalidate()
        }

    private fun extractInitial(letter: String?): String {
        return if (letter == null || letter.trim { it <= ' ' }.isEmpty()) "?" else letter[0].toString()
    }

    private fun updateTextBounds() {
        mTextPaint.getTextBounds(initial, 0, initial.length, mTextBounds)
    }

    companion object {
        const val SHOW_INITIAL = 1
        const val SHOW_IMAGE = 2
        private const val DEF_INITIAL = "A"
        private const val DEF_TEXT_SIZE = 90
        private const val DEF_BACKGROUND_COLOR = 0xE53935

        @State
        private val DEF_STATE = SHOW_INITIAL
    }

    init {
        var text = DEF_INITIAL
        var textColor = Color.WHITE
        var textSize = DEF_TEXT_SIZE
        var backgroundColor = DEF_BACKGROUND_COLOR
        var showState = DEF_STATE
        attrs?.let {
            val a =
                context.obtainStyledAttributes(attrs, R.styleable.AvatarImageView, 0, 0)
            text = a.getString(R.styleable.AvatarImageView_text) ?: ""
            textColor = a.getColor(R.styleable.AvatarImageView_textColor, textColor)
            textSize = a.getDimensionPixelSize(R.styleable.AvatarImageView_textSize, textSize)
            backgroundColor =
                a.getColor(R.styleable.AvatarImageView_avatarBackgroundColor, backgroundColor)
            showState = a.getInt(R.styleable.AvatarImageView_view_state, showState)
            a.recycle()
        }
        mShowState = showState
        mTextPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        mTextPaint.textAlign = Paint.Align.CENTER
        mTextPaint.color = textColor
        mTextPaint.textSize = textSize.toFloat()
        mTextBounds = Rect()
        mText = text
        initial = extractInitial(text)
        updateTextBounds()
        mBackgroundPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        mBackgroundPaint.color = backgroundColor
        mBackgroundPaint.style = Paint.Style.FILL
        mBackgroundBounds = RectF()
    }
}